core/hashbuffer.go

```diff
@@ -0,0 +1,147 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package core
+
+import (
+	"bytes"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// For EVM execution, we need around 256 items. We add a few more to allow reorgs.
+// For mainnet, a couple more would suffice, but a few more added for
+// testnets/private nets
+const hashBufferElems = 288
+
+// hashBuffer implements a storage for chains of hashes, intended to be used for quick lookup of block hashes.
+// Internally, it uses an array of 8192 bytes to store 256 hashes in a circular buffer.
+// It enforces that all hashes added have a contiguous parent-child relation, and supports rollbacks
+// It is not thread-safe.
+type hashBuffer struct {
+	// The data holds the hashes. The hashes are sequential, but also a
+	// circular buffer.
+	// The `head` points to the position of the latest hash.
+	// The parent, if present, is located 32 bytes back.
+	// [.., .., ..., head-2 , head-1, head, oldest, ... ]
+	data [hashBufferElems]common.Hash
+
+	head uint64 // index of hash for the head block
+
+	headNumber uint64 // The block number for head (the most recent block)
+	tailNumber uint64 // The block number for tail (the oldest block)
+}
+
+// Create a new storage with a header in it.
+// Since we take a header here, a hash storage can never be empty.
+// This makes things easier later on (in Set)
+func NewHashStorage(header *types.Header) *hashBuffer {
+	hs := &hashBuffer{}
+	hs.head = 0
+	hs.headNumber = header.Number.Uint64()
+	hs.tailNumber = hs.headNumber
+	hash := header.Hash()
+	copy(hs.data[0][:], hash[:])
+	return hs
+}
+
+// Get locates the hash for the requested number, and copies the data into the
+// provided big.Int -- zeroing it in case the hash is not found
+func (hs *hashBuffer) Get(number uint64) (common.Hash, bool) {
+	var h common.Hash
+	if !hs.has(number) {
+		return h, false
+	}
+	distance := hs.headNumber - number
+	index := (hs.head + hashBufferElems - distance) % hashBufferElems
+	copy(h[:], hs.data[index][:])
+	return h, true
+}
+
+// has returns if the storage has a hash for the given number
+func (hs *hashBuffer) has(number uint64) bool {
+	return number <= hs.headNumber && number >= hs.tailNumber
+}
+
+// Contains checks if the hash at the given number matches the expected
+func (hs *hashBuffer) Contains(number uint64, expected common.Hash) bool {
+	if !hs.has(number) {
+		return false
+	}
+	distance := hs.headNumber - number
+	index := (hs.head + hashBufferElems - distance) % hashBufferElems
+	return bytes.Compare(hs.data[index][:], expected[:]) == 0
```

Comment:
You can compare two hashes, no need for bytes extraction and byte-by-byte comparison.